File: ws01_plumbing/src/py06_time/py06_time/demo01_time.py


import rclpy
from rclpy.node import Node
import time
import threading

from rclpy.time import Time, Duration

class MyNode(Node):
    def __init__(self):
        super().__init__("time_node_py")
        # self.demo_rate()
        # self.demo_time()
        # self.demo_duration()
        self.demo_opt()

    def demo_rate(self):
        self.rate = self.create_rate(1.0)

        thread = threading.Thread(target=self.do_some)
        thread.start()

    def do_some(self):
        while rclpy.ok():
            self.get_logger().info("hello")
            self.rate.sleep()             # 会导致程序阻塞，不能放在主线程
            # time.sleep(1)
    
    def demo_time(self):
        self.time1 = Time(seconds=5, nanoseconds=500000000)
        self.time2 = self.get_clock().now()

        self.get_logger().info("s = %.2f, ns = %d" % (self.time1.seconds_nanoseconds()[0], self.time1.seconds_nanoseconds()[1]))
        self.get_logger().info("s = %.2f, ns = %d" % (self.time2.seconds_nanoseconds()[0], self.time2.seconds_nanoseconds()[1]))

    def demo_duration(self):
        du1 = Duration(seconds=1, nanoseconds=50000000)
        self.get_logger().info("ns = %d" % du1.nanoseconds)

    def demo_opt(self):
        t1 = Time(seconds=20)
        t2 = Time(seconds=15)

        d1 = Duration(seconds=7)
        d2 = Duration(seconds=13)


def main():
    rclpy.init()
    rclpy.spin(MyNode())
    rclpy.shutdown()



if __name__ == '__main__':
    main()

File: ws01_plumbing/src/cpp06_time/src/demo01_time.cpp

#include "rclcpp/rclcpp.hpp"


using namespace std::chrono_literals;

class TimeNode: public rclcpp::Node{
public:
    TimeNode():Node("time_node_cpp"){
        // demo_rate();
        // demo_time();
        // demo_duration();
        demo_opt();
    }


private:
    void demo_rate(){
        // 创建rate对象
        rclcpp::Rate rate1(1s);
        rclcpp::Rate rate2(1.0);
        //调用sleep
        while (rclcpp::ok())
        {
            RCLCPP_INFO(this->get_logger(), "-----------");
            // rate1.sleep();
            rate2.sleep();
        }
        
        
    }

    void demo_time(){
        // 创建time对象
        rclcpp::Time time1(500000000L);
        rclcpp::Time time2(2, 500000000L);
        rclcpp::Time right_now = this->get_clock()->now();
        rclcpp::Time right_now2 = this->now();


        RCLCPP_INFO(this->get_logger(), "s = %.2f, ns = %ld", time1.seconds(), time1.nanoseconds());
        RCLCPP_INFO(this->get_logger(), "s = %.2f, ns = %ld", time2.seconds(), time2.nanoseconds());
        RCLCPP_INFO(this->get_logger(), "s = %.2f, ns = %ld", right_now2.seconds(), right_now2.nanoseconds());

    }

    void demo_duration(){
      rclcpp::Duration du1(1s);
      rclcpp::Duration du2(1, 500000000);

      RCLCPP_INFO(this->get_logger(), "s = %.2f, ns = %ld", du1.seconds(), du1.nanoseconds());
      RCLCPP_INFO(this->get_logger(), "s = %.2f, ns = %ld", du2.seconds(), du2.nanoseconds());
    }

    void demo_opt(){
      // 演示运算法的使用
      rclcpp::Time t1(10, 0);
      rclcpp::Time t2(30, 0);

      rclcpp::Duration d1(10, 0);
      rclcpp::Duration d2(30, 0);

      RCLCPP_INFO(this->get_logger(), "t1 >= t2 ? %d", t1 >= t2);
      RCLCPP_INFO(this->get_logger(), "t1 < t2 ? %d", t1 < t2);
      rclcpp::Duration d3 = t2 - t1;
      rclcpp::Time t3 = t1 + d1;
      rclcpp::Time t4 = t1 - d1;

      RCLCPP_INFO(this->get_logger(), "d3 = %2f", d3.seconds());  // 20
      RCLCPP_INFO(this->get_logger(), "t3 = %2f", t3.seconds());  // 20
      RCLCPP_INFO(this->get_logger(), "t4 = %2f", t4.seconds());  // 0

      RCLCPP_INFO(this->get_logger(), "d1 >= d2 ? %d", d1 >= d2);
      RCLCPP_INFO(this->get_logger(), "d1 < d2 ? %d", d1 < d2);
      rclcpp::Duration d4 = d1 * 3;
      rclcpp::Duration d5 = d2 - d1;
      rclcpp::Duration d6 = d1 + d2;
      RCLCPP_INFO(this->get_logger(), "d4 = %2f", d4.seconds());  // 30
      RCLCPP_INFO(this->get_logger(), "d5 = %2f", d5.seconds());  // 30
      RCLCPP_INFO(this->get_logger(), "d6 = %2f", d6.seconds());  // 30

    }
};




int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TimeNode>());
    rclcpp::shutdown();
    return 0;
}